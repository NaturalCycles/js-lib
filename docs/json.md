# Json

## \_jsonParseIfPossible

Attempts to parse object as JSON.

Returns original object if JSON parse failed (silently).

```ts
_jsonParseIfPossible('abc') // 'abc' (no change, not a json string)
_jsonParseIfPossible(null) // null (no change)
_jsonParseIfPossible({ a: 'a' }) // {a: 'a'} (same object, not a json string)
_jsonParseIfPossible('{"a": "a"}') // {a: 'a'} gotcha! parsed json string into an object!
```

## \_stringify

Inspired by `_inspect` from [nodejs-lib](https://github.com/NaturalCycles/nodejs-lib), which is
based on `util.inpect` that is not available in the Browser.

Transforms `any` to human-readable string (via `JSON.stringify` pretty).

Safe (no error throwing).

Correclty prints `Error`, `AppError`, `ErrorObject`: `error.message + '\n' + _stringify(error.data)`

Enforces max length (default to `1000`, pass `0` to skip it).

Logs numbers as-is (as a String), e.g: `6`.

Logs strings as-is (without single quotes around, unlike default util.inspect behavior).

Otherwise - just uses `JSON.stringify`.

Returns `empty_string` string if empty string is passed.

Returns `undefined` (not a string, but actual `undefined`) if `undefined` is passed (default
`util.inspect` behavior).

```ts
_stringify(undefined) // 'undefined'
_stringify(null) // 'null'
_stringify(true) // 'true'
_stringify(false) // 'false'
_stringify(NaN) // 'null'
_stringify(Infinity) // 'null'
_stringify('') // 'empty_string'
_stringify(' ') // 'empty_string'
_stringify('ho ho ho') // 'ho ho ho'
_stringify(15) // '15'
_stringify(new Error('some msg')) // 'Error: some msg'

// AppError is stringified with it's Data object
_stringify(new AppError('some msg', { k1: 'v1' }))
// 'AppError: some msg\n
// {
//   "k1": "v1"
// }'
```
