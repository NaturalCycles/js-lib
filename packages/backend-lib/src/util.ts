const { GAE_INSTANCE, K_SERVICE } = process.env

export function isGAE(): boolean {
  return !!GAE_INSTANCE
}

export function isCloudRun(): boolean {
  return !!K_SERVICE
}
