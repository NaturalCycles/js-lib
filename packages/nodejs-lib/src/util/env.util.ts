import 'dotenv/config' // ensure .env is read before requiring keys
import type { ValuesOf } from '@naturalcycles/js-lib'
import { fs2 } from '../fs/fs2.js'

/**
 * @example
 *
 * const {a, b} = requreEnvKeys(['a', 'b'])
 *
 * Will throw if any of the passed keys is not defined.
 */
export function requireEnvKeys<T extends readonly string[]>(
  ...keys: T
): { [k in ValuesOf<T>]: string } {
  // eslint-disable-next-line unicorn/no-array-reduce
  return keys.reduce(
    (r, k) => {
      const v = process.env[k]
      if (!v) throw new Error(`${k} env variable is required, but missing`)
      r[k as ValuesOf<T>] = v
      return r
    },
    {} as { [k in ValuesOf<T>]: string },
  )
}

/**
 * @deprecated use fs2.requireFileToExist
 */
export function requireFileToExist(filePath: string): void {
  fs2.requireFileToExist(filePath)
}
