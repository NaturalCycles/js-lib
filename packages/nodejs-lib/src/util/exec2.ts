import { execSync, spawn, spawnSync } from 'node:child_process'
import {
  _substringAfterLast,
  type AnyObject,
  type NumberOfMilliseconds,
  type UnixTimestampMillis,
} from '@naturalcycles/js-lib'
import { _since, AppError } from '@naturalcycles/js-lib'
import { dimGrey, dimRed, hasColors, white } from '../colors/colors.js'

/**
 * Set of utility functions to work with Spawn / Exec.
 *
 * How to decide between Spawn and Exec?
 *
 * Long-running job that prints output, and no need to return the output - use Spawn.
 *
 * Short-running job, no need to print the output, might want to return the output - use Exec.
 *
 * Need to both print and return the output - use SpawnAsyncAndReturn.
 *
 * ***
 *
 * Spawn is good for long-running large-output processes, that continuously output data.
 * E.g running `jest`.
 *
 * Exec is the opposite - good for short-running processes that output small data.
 * Exec allows to return the output as a string.
 * Exec doesn't stream data during execution, so the output/error will only be printed
 * at the end.
 * Exec always uses the shell (there's no option to disable it).
 */
class Exec2 {
  /**
   * Reasons to use it:
   * - Sync
   * - Need to print output while running
   *
   * Limitations:
   * - Cannot return stdout/stderr (use exec, execAsync or spawnAsyncAndReturn for that)
   *
   * Defaults:
   *
   * shell: true
   * log: true
   */
  spawn(cmd: string, opt: SpawnOptions = {}): void {
    const { shell = true, cwd, env, passProcessEnv = true, forceColor = hasColors } = opt
    opt.log ??= true // by default log should be true, as we are printing the output
    opt.logStart ??= opt.log
    opt.logFinish ??= opt.log
    const started = Date.now() as UnixTimestampMillis
    this.logStart(cmd, opt)

    const r = spawnSync(cmd, opt.args, {
      encoding: 'utf8',
      stdio: 'inherit',
      shell,
      cwd,
      env: {
        ...(passProcessEnv ? process.env : {}),
        ...(forceColor ? { FORCE_COLOR: '1' } : {}),
        ...env,
      },
    })

    const isSuccessful = !r.error && !r.status
    this.logFinish(cmd, opt, started, isSuccessful)

    if (r.error) {
      throw r.error
    }
    if (r.status) {
      throw new Error(`spawn exited with code ${r.status}: ${cmd}`)
    }
  }

  /**
   * Reasons to use it:
   *
   * - Sync
   * - Need to return output
   *
   * Limitations:
   * - Cannot print while running (use spawn or spawnAsync for that)
   *
   * Defaults:
   *
   * shell: true
   * log: false
   */
  exec(cmd: string, opt: ExecOptions = {}): string {
    const { cwd, env, passProcessEnv = true, timeout } = opt
    opt.logStart ??= opt.log ?? false
    opt.logFinish ??= opt.log ?? false
    const started = Date.now() as UnixTimestampMillis
    this.logStart(cmd, opt)

    try {
      const s = execSync(cmd, {
        encoding: 'utf8',
        // stdio: 'inherit', // no, otherwise we don't get the output returned
        stdio: undefined,
        // shell: undefined,
        cwd,
        timeout,
        env: {
          ...(passProcessEnv ? process.env : {}),
          ...env,
        },
      }).trim()

      this.logFinish(cmd, opt, started, true)
      return s
    } catch (err) {
      // Not logging stderr, as it's printed by execSync by default (somehow)
      // stdout is not printed by execSync though, therefor we print it here
      // if ((err as any).stderr) {
      //   process.stderr.write((err as any).stderr)
      // }
      if ((err as any).stdout) {
        process.stdout.write((err as any).stdout)
      }
      this.logFinish(cmd, opt, started, false)
      throw new Error(`exec exited with code ${(err as any).status}: ${cmd}`)
    }
  }

  /**
   * Reasons to use it:
   * - Async
   * - Need to print output while running
   *
   * Limitations:
   * - Cannot return stdout/stderr (use execAsync or spawnAsyncAndReturn for that)
   *
   * Defaults:
   *
   * shell: true
   * log: true
   */
  async spawnAsync(cmd: string, opt: SpawnOptions = {}): Promise<void> {
    const { shell = true, cwd, env, passProcessEnv = true, forceColor = hasColors } = opt
    opt.log ??= true // by default log should be true, as we are printing the output
    opt.logStart ??= opt.log
    opt.logFinish ??= opt.log
    const started = Date.now() as UnixTimestampMillis
    this.logStart(cmd, opt)

    await new Promise<void>((resolve, reject) => {
      const p = spawn(cmd, opt.args || [], {
        shell,
        cwd,
        stdio: 'inherit',
        env: {
          ...(passProcessEnv ? process.env : {}),
          ...(forceColor ? { FORCE_COLOR: '1' } : {}),
          ...env,
        },
      })

      p.on('close', code => {
        const isSuccessful = !code
        this.logFinish(cmd, opt, started, isSuccessful)
        if (code) {
          return reject(new Error(`spawnAsync exited with code ${code}: ${cmd}`))
        }
        resolve()
      })
    })
  }

  /**
   * Advanced/async version of Spawn.
   * Consider simpler `spawn` or `exec` first, which are also sync.
   *
   * spawnAsyncAndReturn features:
   *
   * 1. Async
   * 2. Allows to collect the output AND print it while running.
   * 3. Returns SpawnOutput with stdout, stderr and exitCode.
   * 4. Allows to not throw on error, but just return SpawnOutput for further inspection.
   *
   * Defaults:
   *
   * shell: true
   * printWhileRunning: true
   * collectOutputWhileRunning: true
   * throwOnNonZeroCode: true
   * log: true
   */
  async spawnAsyncAndReturn(cmd: string, opt: SpawnAsyncOptions = {}): Promise<SpawnOutput> {
    const {
      shell = true,
      printWhileRunning = true,
      collectOutputWhileRunning = true,
      throwOnNonZeroCode = true,
      cwd,
      env,
      passProcessEnv = true,
      forceColor = hasColors,
    } = opt
    opt.log ??= printWhileRunning // by default log should be true, as we are printing the output
    opt.logStart ??= opt.log
    opt.logFinish ??= opt.log
    const started = Date.now() as UnixTimestampMillis
    this.logStart(cmd, opt)
    let stdout = ''
    let stderr = ''

    return await new Promise<SpawnOutput>((resolve, reject) => {
      const p = spawn(cmd, opt.args || [], {
        shell,
        cwd,
        env: {
          ...(passProcessEnv ? process.env : {}),
          ...(forceColor ? { FORCE_COLOR: '1' } : {}),
          ...env,
        },
      })

      p.stdout.on('data', data => {
        if (collectOutputWhileRunning) {
          stdout += data.toString()
          // console.log('stdout:', data.toString())
        }
        if (printWhileRunning) {
          process.stdout.write(data)
          // console.log('stderr:', data.toString())
        }
      })
      p.stderr.on('data', data => {
        if (collectOutputWhileRunning) {
          stderr += data.toString()
        }
        if (printWhileRunning) {
          process.stderr.write(data)
        }
      })

      p.on('close', code => {
        const isSuccessful = !code
        this.logFinish(cmd, opt, started, isSuccessful)
        const exitCode = code || 0
        const o: SpawnOutput = {
          exitCode,
          stdout: stdout.trim(),
          stderr: stderr.trim(),
        }
        if (throwOnNonZeroCode && code) {
          return reject(new SpawnError(`spawnAsyncAndReturn exited with code ${code}: ${cmd}`, o))
        }
        resolve(o)
      })
    })
  }

  private logStart(cmd: string, opt: SpawnOptions | ExecOptions): void {
    if (!opt.logStart) return

    console.log(
      [
        dimGrey(...Object.entries(opt.env || {}).map(([k, v]) => [k, v].join('='))),
        white(opt.name || _substringAfterLast(cmd, '/')),
        ...((!opt.name && (opt as SpawnOptions).args) || []),
      ]
        .filter(Boolean)
        .join(' '),
    )
  }

  private logFinish(
    cmd: string,
    opt: SpawnOptions | ExecOptions,
    started: UnixTimestampMillis,
    isSuccessful: boolean,
  ): void {
    if (isSuccessful && !opt.logFinish) return

    console.log(
      [
        white(opt.name || _substringAfterLast(cmd, '/')),
        ...((!opt.name && (opt as SpawnOptions).args) || []),
        dimGrey('took ' + _since(started)),
        !isSuccessful && dimGrey('and ') + dimRed('failed'),
      ]
        .filter(Boolean)
        .join(' '),
    )
  }
}

export const exec2 = new Exec2()

export class SpawnError extends AppError<SpawnErrorData> {
  constructor(message: string, data: SpawnErrorData) {
    super(message, data, { name: 'SpawnError' })
  }
}

export interface SpawnErrorData extends SpawnOutput {}

export interface SpawnOutput {
  /**
   * Exit code of the spawned process.
   * 0 means success, anything else means failure.
   */
  exitCode: number
  stdout: string
  stderr: string
}

export interface SpawnAsyncOptions extends SpawnOptions {
  /**
   * Defaults to true.
   * If true - prints both stdout and stderr to console while running,
   * otherwise runs "silently".
   * Returns SpawnOutput in the same way, regardless of `printWhileRunning` setting.
   */
  printWhileRunning?: boolean

  /**
   * Defaults to true.
   * If true - collects stdout and stderr while running, and return it in the end.
   * stdout/stderr are collected and returned regardless if it returns with error or not.
   * On success - stdout/stderr are available from `SpawnOutput`.
   * On error - stdout/stderr are available from `SpawnError.data`.
   */
  collectOutputWhileRunning?: boolean

  /**
   * Defaults to true.
   * If true - throws SpawnError if non-zero code is returned.
   * SpawnError conveniently contains .data.stdout and .data.strerr for inspection.
   * If false - will not throw, but return SpawnOutput with stdout, stderr and exitCode.
   */
  throwOnNonZeroCode?: boolean
}

export interface SpawnOptions {
  args?: string[]
  /**
   * Defaults to true.
   */
  logStart?: boolean
  /**
   * Defaults to true.
   */
  logFinish?: boolean
  /**
   * Defaults to true.
   * Controls/overrides both logStart and logFinish simultaneously.
   */
  log?: boolean
  /**
   * Defaults to true.
   */
  shell?: boolean

  /**
   * If specified - will be used as "command name" for logging purposes,
   * instead of "cmd + args"
   */
  name?: string
  cwd?: string

  env?: AnyObject
  /**
   * Defaults to true.
   * Set to false to NOT pass `process.env` to the spawned process.
   */
  passProcessEnv?: boolean
  /**
   * Defaults to "auto detect colors".
   * Set to false or true to override.
   */
  forceColor?: boolean
}

export interface ExecOptions {
  /**
   * Defaults to false.
   */
  logStart?: boolean
  /**
   * Defaults to false.
   */
  logFinish?: boolean
  /**
   * Defaults to false.
   * Controls/overrides both logStart and logFinish simultaneously.
   */
  log?: boolean

  /**
   * If specified - will be used as "command name" for logging purposes,
   * instead of "cmd + args"
   */
  name?: string
  cwd?: string
  timeout?: NumberOfMilliseconds

  env?: AnyObject
  /**
   * Defaults to false for security reasons.
   * Set to true to pass `process.env` to the spawned process.
   */
  passProcessEnv?: boolean
}
