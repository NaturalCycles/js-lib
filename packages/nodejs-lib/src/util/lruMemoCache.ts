import type { MemoCache } from '@naturalcycles/js-lib'
import { LRUCache } from 'lru-cache'

// Partial, to be able to provide default `max`
export type LRUMemoCacheOptions<KEY, VALUE> = Partial<LRUCache.Options<KEY, VALUE, any>>

/**
 * @example
 * Use it like this:
 *
 * @_Memo({ cacheFactory: () => new LRUMemoCache({...}) })
 * method1 ()
 */
export class LRUMemoCache<KEY = any, VALUE = any> implements MemoCache<KEY, VALUE> {
  constructor(opt: LRUMemoCacheOptions<KEY, VALUE>) {
    this.lru = new LRUCache<any, any>({
      max: 100,
      ...opt,
    })
  }

  private lru!: LRUCache<any, any>

  has(k: any): boolean {
    return this.lru.has(k)
  }

  get(k: any): any {
    return this.lru.get(k)
  }

  set(k: any, v: any): void {
    this.lru.set(k, v)
  }

  clear(): void {
    this.lru.clear()
  }
}
