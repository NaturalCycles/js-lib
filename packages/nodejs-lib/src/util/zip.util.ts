import { promisify } from 'node:util'
import type { ZlibOptions } from 'node:zlib'
import zlib from 'node:zlib'

const deflate = promisify(zlib.deflate.bind(zlib))
const inflate = promisify(zlib.inflate.bind(zlib))
const gzip = promisify(zlib.gzip.bind(zlib))
const gunzip = promisify(zlib.gunzip.bind(zlib))

// string > compressed buffer

/**
 * deflateBuffer uses `deflate`.
 * It's 9 bytes shorter than `gzip`.
 */
export async function deflateBuffer(buf: Buffer, options: ZlibOptions = {}): Promise<Buffer> {
  return await deflate(buf, options)
}

export async function inflateBuffer(buf: Buffer, options: ZlibOptions = {}): Promise<Buffer> {
  return await inflate(buf, options)
}

/**
 * deflateString uses `deflate`.
 * It's 9 bytes shorter than `gzip`.
 */
export async function deflateString(s: string, options?: ZlibOptions): Promise<Buffer> {
  return await deflateBuffer(Buffer.from(s), options)
}

export async function inflateToString(buf: Buffer, options?: ZlibOptions): Promise<string> {
  return (await inflateBuffer(buf, options)).toString()
}

/**
 * gzipBuffer uses `gzip`
 * It's 9 bytes longer than `deflate`.
 */
export async function gzipBuffer(buf: Buffer, options: ZlibOptions = {}): Promise<Buffer> {
  return await gzip(buf, options)
}

export async function gunzipBuffer(buf: Buffer, options: ZlibOptions = {}): Promise<Buffer> {
  return await gunzip(buf, options)
}

/**
 * gzipString uses `gzip`.
 * It's 9 bytes longer than `deflate`.
 */
export async function gzipString(s: string, options?: ZlibOptions): Promise<Buffer> {
  return await gzipBuffer(Buffer.from(s), options)
}

export async function gunzipToString(buf: Buffer, options?: ZlibOptions): Promise<string> {
  return (await gunzipBuffer(buf, options)).toString()
}
