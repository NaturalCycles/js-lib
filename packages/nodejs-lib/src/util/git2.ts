import { execSync } from 'node:child_process'
import { basename } from 'node:path'
import type { UnixTimestamp } from '@naturalcycles/js-lib'
import { grey } from '../colors/colors.js'
import { exec2 } from './exec2.js'

/**
 * Set of utility functions to work with git.
 */
class Git2 {
  getLastGitCommitMsg(): string {
    return exec2.exec('git log -1 --pretty=%B')
  }

  commitMessageToTitleMessage(msg: string): string {
    const firstLine = msg.split('\n')[0]!
    const [preTitle, title] = firstLine.split(': ')
    return title || preTitle!
  }

  hasUncommittedChanges(): boolean {
    // git diff-index --quiet HEAD -- || echo "untracked"
    try {
      execSync('git diff-index --quiet HEAD --', {
        encoding: 'utf8',
      })
      return false
    } catch {
      return true
    }
  }

  /**
   * Returns true if there were changes
   */
  commitAll(msg: string): boolean {
    // git commit -a -m "style(lint-all): $GIT_MSG" || true
    const cmd = `git commit -a --no-verify -m "${msg}"`
    // const cmd = `git`
    // const args = ['commit', '-a', '--no-verify', '-m', msg]
    console.log(grey(cmd))

    try {
      execSync(cmd, {
        stdio: 'inherit',
      })
      return true
    } catch {
      return false
    }
  }

  /**
   * @returns true if there are not pushed commits.
   */
  isAhead(): boolean {
    // ahead=`git rev-list HEAD --not --remotes | wc -l | awk '{print $1}'`
    const cmd = `git rev-list HEAD --not --remotes | wc -l | awk '{print $1}'`
    const stdout = exec2.exec(cmd)
    // console.log(`gitIsAhead: ${stdout}`)
    return Number(stdout) > 0
  }

  pull(): void {
    const cmd = 'git pull'
    try {
      execSync(cmd, {
        stdio: 'inherit',
      })
    } catch {}
  }

  push(): void {
    // git push --set-upstream origin $CIRCLE_BRANCH && echo "pushed, exiting" && exit 0
    let cmd = 'git push'

    const branchName = this.getCurrentBranchName()

    if (branchName) {
      cmd += ` --set-upstream origin ${branchName}`
    }

    exec2.spawn(cmd, { logStart: true })
  }

  getCurrentCommitSha(full = false): string {
    const sha = exec2.exec('git rev-parse HEAD')
    return full ? sha : sha.slice(0, 7)
  }

  getCurrentCommitTimestamp(): UnixTimestamp {
    return Number(exec2.exec('git log -1 --format=%ct')) as UnixTimestamp
  }

  getCurrentBranchName(): string {
    return exec2.exec('git rev-parse --abbrev-ref HEAD')
  }

  getCurrentRepoName(): string {
    const originUrl = exec2.exec('git config --get remote.origin.url')
    return basename(originUrl, '.git')
  }
}

export const git2 = new Git2()
