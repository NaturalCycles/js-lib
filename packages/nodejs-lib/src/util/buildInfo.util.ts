import type { AnyObject, BuildInfo, UnixTimestamp } from '@naturalcycles/js-lib'
import { _filterUndefinedValues, localTime } from '@naturalcycles/js-lib'
import { fs2 } from '../fs/fs2.js'
import { git2 } from './git2.js'

export interface GenerateBuildInfoOptions {
  /**
   * If set - this timestamp will be used, instead of "current time".
   */
  overrideTimestamp?: UnixTimestamp

  /**
   * Defaults to currently checked out git branch.
   */
  overrideBranchName?: string
}

export function generateBuildInfo(opt: GenerateBuildInfoOptions = {}): BuildInfo {
  const now = localTime.orNow(opt.overrideTimestamp)
  const ts = now.unix

  const rev = git2.getCurrentCommitSha()
  const branchName = opt.overrideBranchName || git2.getCurrentBranchName()
  const repoName = git2.getCurrentRepoName()
  const tsCommit = git2.getCurrentCommitTimestamp()

  const ver = [now.toStringCompact(), repoName, branchName, rev].join('_')

  let { APP_ENV: env } = process.env

  if (!env) {
    // Attempt to read `envByBranch` from package.json root
    try {
      if (fs2.pathExists('package.json')) {
        const packageJson = fs2.readJson<AnyObject>('package.json')
        env = packageJson?.['envByBranch']?.[branchName] || packageJson?.['envByBranch']?.['*']
      }
    } catch {}
  }

  return _filterUndefinedValues({
    ts,
    tsCommit,
    repoName,
    branchName,
    rev,
    ver,
    env,
  })
}
