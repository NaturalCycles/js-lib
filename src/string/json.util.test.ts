import { expectResults, mockAllKindsOfThings } from '@naturalcycles/dev-lib/dist/testing'
import { _jsonParseIfPossible } from './json.util'

test('jsonParseIfPossible', () => {
  expectResults(v => _jsonParseIfPossible(v), mockAllKindsOfThings()).toMatchSnapshot()

  expect(_jsonParseIfPossible('{"a": "b"}')).toEqual({ a: 'b' })
  expect(_jsonParseIfPossible('  {"a": "b"}')).toEqual({ a: 'b' })
  expect(_jsonParseIfPossible('5')).toBe(5)
  expect(_jsonParseIfPossible('something')).toBe('something')
  expect(_jsonParseIfPossible('{something')).toBe('{something')
})
