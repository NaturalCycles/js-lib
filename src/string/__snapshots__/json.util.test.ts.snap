// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsonParseIfPossible 1`] = `
Map {
  undefined => undefined,
  null => null,
  true => true,
  false => false,
  NaN => NaN,
  Infinity => Infinity,
  "" => "",
  " " => " ",
  "ho ho ho" => "ho ho ho",
  "{" => "{",
  "{"a": "a1"}" => {
    "a": "a1",
  },
  "[5, null, "s"]" => [
    5,
    null,
    "s",
  ],
  [
    5,
    null,
    "s",
  ] => [
    5,
    null,
    "s",
  ],
  15 => 15,
  [
    1,
    2,
    3,
  ] => [
    1,
    2,
    3,
  ],
  {
    "a": "a1",
    "b": "b1",
    "c": {
      "d": 25,
      "e": undefined,
    },
  } => {
    "a": "a1",
    "b": "b1",
    "c": {
      "d": 25,
      "e": undefined,
    },
  },
  {
    "documentation_url": "https://developer.github.com/v3/git/refs/#create-a-reference",
    "message": "Reference already exists",
  } => {
    "documentation_url": "https://developer.github.com/v3/git/refs/#create-a-reference",
    "message": "Reference already exists",
  },
  [Function] => [Function],
  /i am regex, who are you\\?/ => /i am regex, who are you\\?/,
  Map {
    "a" => "b",
  } => Map {
    "a" => "b",
  },
  Set {
    "a",
    "b",
    "c",
  } => Set {
    "a",
    "b",
    "c",
  },
  [Error: some err msg] => [Error: some err msg],
  [Error: Error with code] => [Error: Error with code],
  [AppError: app error] => [AppError: app error],
  Symbol(a_am_symbol) => Symbol(a_am_symbol),
  1984-06-21T00:00:00.000Z => 1984-06-21T00:00:00.000Z,
}
`;
